Add wasm.cmake for best practices to image

alfgrini75 opened the pull request:
This allows the different repos which use this image of the image to use a "gold standard" cmake.wasm file from within this repo instead of relying on their own wasm.cmake file (or maybe relying on one "gold standard repo" which they can copy a file from).

A common use case is that one project uses a stack of different projects or libs. With this change, the user can include in their CMakeLists.txt this file to get consistent build settings

They would then just include include(/opt/wasm-deps/share/cmake/wasm.cmake) in their cmake lists

## Existing /opt/wasm-deps/share content
```
# ls -l /opt/wasm-deps/share
total 44
-rw-r--r-- 1 root root  5517 Oct  9 13:09 COPYING
-rw-r--r-- 1 root root 13712 Oct  9 13:09 RELEASE.txt
-rw-r--r-- 1 root root 10682 Oct  9 13:09 USING_HDF5_CMake.txt
drwxr-xr-x 3 root root  4096 Oct  9 13:12 eigen3
drwxr-xr-x 2 root root  4096 Oct  9 13:12 pkgconfig

```

forderud commented on image/wasm.cmake:
```diff
@@ -0,0 +1,21 @@
+
+# Enable SSE2 support
+# https://emscripten.org/docs/porting/simd.html
+add_compile_options(-msse2 -mrelaxed-simd)
```
Interesting idea. A different version of `wasm.cmake` is already in the `sample` folder in this repo. I don't want two versions under version control. Therefore, I suggest moving the existing file as-is and update the `sample` project to still build as before.

I suggest discussing build setting changes in separate pull requests.

forderud commented on image/wasm.cmake:
```diff
@@ -0,0 +1,21 @@
+# Enable only WebGL 2
+# https://emscripten.org/docs/porting/multimedia_and_graphics/OpenGL-support.html
+add_link_options("SHELL:-s MAX_WEBGL_VERSION=2")
+add_link_options("SHELL:-s MIN_WEBGL_VERSION=2")
```
If you take a look at [wasm.conf](https://github.com/qt/qtbase/blob/dev/mkspecs/common/wasm/wasm.conf) and [QtWasmHelpers.cmake](https://github.com/qt/qtbase/blob/dev/cmake/QtWasmHelpers.cmake) in the Qt "qtbase" sources then you'll see that Qt is already defining `MAX_WEBGL_VERSION=2`.

Are you therefore sure that it's needed to also set `MAX_WEBGL_VERSION=2` here? If so, then please propose it in a separate PR with some form of justification.

alfgrini75 replied:
Ok, I change this MR to ONLY be a "common file which is part of the image". Let's discuss any change of build options later.

Commit:
Move the wasm.cmake used by the sample app to be part of the image

The idea is that any other app building with this image can use the
same wasm.cmake and therefore use some common "best practices" build options
when using this image

forderud commented on image/Dockerfile:
```diff
@@ -94,6 +94,10 @@ FROM emscripten_base
 RUN update-ca-certificates
+
+# Copy the wasm.cmake build configuration file
+COPY wasm.cmake /opt/wasm-deps/share/cmake/
```
Is the the `/share/cmake/wasm.cmake` naming convention here a well-established pattern also found elsewhere, or just your best suggestion for a suitable location?

alfgrini75 replied:
This location was a suggestion from AI as best practice. I don't have any documentation for the choice of location.

forderud replied:
Ok. I don't have any better suggestions myself, so it's probably a good place for now.

It would be very nice if we could get CMake to auto-detect the `wasm.cmake` file by placing it in a special folder, but we can look into that later.

forderud commented:
@alfgrini75 Can you please confirm that you've tested your PR and inspected the generated image to verify that it contains a new `/opt/wasm-deps/share/cmake/wasm.cmake` file?

You can use `image/build_image.bat` to build a local image for testing. The docker "push" and "tag" steps at the end will fail, but don't worry about them. You can afterwards inspect the generated image with `run_shell.bat`.

alfgrini75 commented:
> @alfgrini75 Can you please confirm that you've tested your PR and inspected the generated image to verify that it contains a new `/opt/wasm-deps/share/cmake/wasm.cmake` file?
> 
> You can use `image/build_image.bat` to build a local image for testing. The docker "push" and "tag" steps at the end will fail, but don't worry about them. You can afterwards inspect the generated image with `run_shell.bat`.

<img width="862" height="361" alt="image" src="https://github.com/user-attachments/assets/f9da88b2-dd68-4a14-b72e-5d78fc20839c" />

forderud commented on sample/CMakeLists.txt:
```diff
@@ -4,7 +4,7 @@ set(CMAKE_CXX_STANDARD 17)
 
 if(EMSCRIPTEN)
-  include(wasm.cmake)
+  include(/opt/wasm-deps/share/cmake/wasm.cmake)
```
This will break the build for existing Docker images. Can you instead change the include to something like `include(../image/wasm.cmake)` to avoid causing near-term breakage?

We can change to your proposed toolchain path at a later point.

alfgrini75 replied:
This is ok for me, but I don't understand it since if building sample app with this version of the image, it should work. But will you run old versions of the sampleapp in this new version of the image?  or new versions of sample app in old images?

forderud replied:
There's not yet any `qtwasm` docker images on hub.docker.com that contains the `wasm.cmake` file introduced by this PR. I don't want to break this sample project until new Docker images are available.

Commit:
Do not use the wasm.cmake from /opt/wasm-deps for the test app

To avoid problems if running a new version of the sampleapp in older images
Note that new versions of the sample app will need to run new versions
of build_cmake.bat or build_cmake.sh

forderud commented on sample/CMakeLists.txt:
```diff
@@ -4,7 +4,7 @@ set(CMAKE_CXX_STANDARD 17)
 
 if(EMSCRIPTEN)
-  include(wasm.cmake)
+  include(../image/wasm.cmake)
```
I get the following error when attempting to run `build_cmake.bat` from your PR:
```
C:\Dev\QtWasm>build_cmake.bat

C:\Dev\QtWasm>rmdir build /s /q

C:\Dev\QtWasm>mkdir build

C:\Dev\QtWasm>docker run --rm -v C:\Dev\QtWasm\sample:/project/source -v C:\Dev\QtWasm\build:/project/build forderud/qtwasm:latest   || exit /b 1
-- The C compiler identification is Clang 21.0.0
-- The CXX compiler identification is Clang 21.0.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /emsdk/upstream/emscripten/emcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /emsdk/upstream/emscripten/em++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:7 (include):
  include could not find requested file:

    ../image/wasm.cmake


-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Using Qt bundled ZLIB.
-- Found WrapZLIB: TRUE (found version "6.10.0")
-- Performing Test HAVE_STDATOMIC
-- Performing Test HAVE_STDATOMIC - Success
-- Found WrapAtomic: TRUE
-- Using Qt bundled PCRE2.
-- Found WrapPCRE2: TRUE (found version "6.10.0")
-- Performing Test HAVE_GETTIME
-- Performing Test HAVE_GETTIME - Success
-- Performing Test HAVE_SHM_OPEN_SHM_UNLINK
-- Performing Test HAVE_SHM_OPEN_SHM_UNLINK - Success
-- Found WrapRt: TRUE
-- Found GLESv2: ON
-- Performing Test HAVE_EGL
-- Performing Test HAVE_EGL - Success
-- Found EGL: /emsdk/upstream/emscripten/cache/sysroot/include (found version "1.5")
-- Using Qt bundled PNG.
-- Found WrapPNG: TRUE (found version "6.10.0")
-- Using Qt bundled Harfbuzz.
-- Found WrapHarfbuzz: TRUE (found version "6.10.0")
-- Using Qt bundled Freetype.
-- Found WrapFreetype: TRUE (found version "6.10.0")
-- Using Qt bundled Jpeg.
-- Found WrapJpeg: TRUE (found version "6.10.0")
Detected locale "C" with character encoding "ANSI_X3.4-1968", which is not UTF-8.
Qt depends on a UTF-8 locale, and has switched to "C.UTF-8" instead.
If this causes problems, reconfigure your locale. See the locale(1) manual
for more information.
CMake Error at app/CMakeLists.txt:24 (find_package):
  Could not find a package configuration file provided by "Eigen3" with any
  of the following names:

    Eigen3Config.cmake
    eigen3-config.cmake

  Add the installation prefix of "Eigen3" to CMAKE_PREFIX_PATH or set
  "Eigen3_DIR" to a directory containing one of the above files.  If "Eigen3"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/project/build/CMakeFiles/CMakeOutput.log".
See also "/project/build/CMakeFiles/CMakeError.log".
```

Seems like there's something wrong with the relative path.

alfgrini75 replied:
I think the reason is that wasm.cmake is no longer part of the sample directory so when you do "docker run --rm -v C:\Dev\QtWasm\sample:/project/source" you are linking all the code in "sample" to /project/source, but since wasm.cmake is no longer in "sample"  it is not part of this. And the only place it can be found is in new location.

<img width="599" height="122" alt="image" src="https://github.com/user-attachments/assets/b4d4b4b6-ca1e-46ec-a027-3094cbc469c4" />

alfgrini75 replied:
So to go ahead with this, we either need to use /opt/wasm/deps/share/cmake/wasm.cmake (like in the first commit here) or duplicate the file if you want to run old version of the sample app in new versions of the image or vice versa.

Old versions of the sampleapp would have the file in it's source, so old versions of the sampleapp in a new image should work.

Old images with new versions of the sampleapp would not work since wasm.cmake is removed from the sampleapp source and it is not part of /opt/wasm/deps in those old images.

New versions going forward with new sampleapp and new version of the image work if changed as in the first commit here.

forderud replied:
Please instead fix the relative path.

You can use `configure_file` to copy the file if required.

forderud commented:
Wouldn't it be simpler to do something like #97?

alfgrini75 commented:
> Wouldn't it be simpler to do something like #97?

It feels a bit strange to maintain some kind of official file which lives under "sample". (It feels better that it lives in "image". And It is slightly more confusing to follow the file traveling from "sample" to "image" and then to "/opt" folder.

But it is ok for me. I agree it serves the same goal. If you prefer this it is ok and we can move the file later if we want to.

forderud commented:
Closing PR in favor of #97.
